testutil/checkers.go

```diff
@@ -43,41 +44,41 @@ func (c *containsChecker) Check(params []interface{}, names []string) (result bo
 			error = fmt.Sprint(v)
 		}
 	}()
-	var haystack interface{} = params[0]
-	var needle interface{} = params[1]
-	switch haystackV := reflect.ValueOf(haystack); haystackV.Kind() {
-	case reflect.Slice, reflect.Array:
-		// Ensure that type of elements in haystack is compatible with needle
-		if needleV := reflect.ValueOf(needle); haystackV.Type().Elem() != needleV.Type() {
-			panic(fmt.Sprintf("haystack contains items of type %s but needle is a %s",
-				haystackV.Type().Elem(), needleV.Type()))
+	var container interface{} = params[0]
+	var elem interface{} = params[1]
+	// Ensure that type of elements in container is compatible with elem
+	switch containerV := reflect.ValueOf(container); containerV.Kind() {
+	case reflect.Slice, reflect.Array, reflect.Map:
+		if elemV := reflect.ValueOf(elem); containerV.Type().Elem() != elemV.Type() {
+			return false, fmt.Sprintf(
+				"container has items of type %s but expected element is a %s",
+				containerV.Type().Elem(), elemV.Type())
 		}
-		for len, i := haystackV.Len(), 0; i < len; i++ {
-			itemV := haystackV.Index(i)
-			if itemV.Interface() == needle {
+	}
+	switch containerV := reflect.ValueOf(container); containerV.Kind() {
+	case reflect.Slice, reflect.Array:
+		for len, i := containerV.Len(), 0; i < len; i++ {
+			itemV := containerV.Index(i)
+			if itemV.Interface() == elem {
 				return true, ""
 			}
 		}
 		return false, ""
 	case reflect.Map:
-		// Ensure that type of elements in haystack is compatible with needle
-		if needleV := reflect.ValueOf(needle); haystackV.Type().Elem() != needleV.Type() {
-			panic(fmt.Sprintf("haystack contains items of type %s but needle is a %s",
-				haystackV.Type().Elem(), needleV.Type()))
-		}
-		for _, keyV := range haystackV.MapKeys() {
-			itemV := haystackV.MapIndex(keyV)
-			if itemV.Interface() == needle {
+		for _, keyV := range containerV.MapKeys() {
+			itemV := containerV.MapIndex(keyV)
+			if itemV.Interface() == elem {
 				return true, ""
 			}
 		}
 		return false, ""
 	case reflect.String:
-		// When haystack is a string, we expect needle to be a string as well
-		needle := params[1].(string)
-		haystack := params[0].(string)
-		return strings.Contains(haystack, needle), ""
+		// When container is a string, we expect elem to be a string as well
+		if reflect.ValueOf(elem).Kind() != reflect.String {
+			return false, fmt.Sprintf("element is a %T but expected a string", elem)
+		}
+		return strings.Contains(container.(string), elem.(string)), ""
```

Comment:
There's a bug here: the _kind_ of elem may be a string without it being of _type_ string. This logic will panic if it gets a, say, MyString type which has string as an underlying type as an argument (both are of _kind_ reflect.String). The right thing to do is provide containerV.String() and elemV.String() to Contains, after the kinds have been confirmed to be strings.